samza-core/src/main/java/org/apache/samza/operators/impl/SendToTableWithUpdateOperatorImpl.java

```diff
@@ -0,0 +1,123 @@
+    extends OperatorImpl<KV<K, UpdateMessage<U, V>>, KV<K, UpdateMessage<U, V>>> {
+  private static final Logger LOG = LoggerFactory.getLogger(SendToTableWithUpdateOperatorImpl.class);
+
+  private final SendToTableWithUpdateOperatorSpec<K, V, U> spec;
+  private final ReadWriteTable<K, V, U> table;
+
+  public SendToTableWithUpdateOperatorImpl(SendToTableWithUpdateOperatorSpec<K, V, U> spec, Context context) {
+    this.spec = spec;
+    this.table = context.getTaskContext().getTable(spec.getTableId());
+    if (context.getTaskContext().getTable(spec.getTableId()) instanceof RemoteTable) {
+      RemoteTable<K, V, U> remoteTable = (RemoteTable<K, V, U>) table;
+      if (remoteTable.getBatchProvider() instanceof CompactBatchProvider) {
+        throw new SamzaException("Batching is not supported with Compact Batches for partial updates");
+      }
+    }
+  }
+
+  @Override
+  protected void handleInit(Context context) {
+  }
+
+  @Override
+  protected CompletionStage<Collection<KV<K, UpdateMessage<U, V>>>> handleMessageAsync(KV<K, UpdateMessage<U, V>> message,
+      MessageCollector collector, TaskCoordinator coordinator) {
+    final UpdateOptions options = spec.getUpdateOptions();
+    final CompletableFuture<Void> updateFuture = table.updateAsync(message.getKey(), message.getValue().getUpdate());
+
+    return updateFuture
+        .handle((result, ex) -> {
+          if (ex == null) {
+            // success, no need to Put a default value
+            return false;
+          } else if (ex.getCause() instanceof RecordNotFoundException && message.getValue().hasDefault()) {
+            // If an update fails for a given key due to a RecordDoesNotExistException exception thrown and
+            // a default is provided and the UpdateOptions is set to UPDATE_WITH_DEFAULTS, then attempt
+            // to PUT a default record for the key and then apply the update.
+            if (options == UpdateOptions.UPDATE_WITH_DEFAULTS) {
+              return true;
+            } else {
+              throw new SamzaException("Put default failed for update as the Update options was set to " + options +
+                  ". Please use UpdateOptions.UPDATE_WITH_DEFAULTS instead.");
+            }
+          } else {
+            throw new SamzaException("Update failed with exception: ", ex);
+          }
+        })
+        .thenCompose(shouldPutDefault -> {
+          if (shouldPutDefault) {
+            final CompletableFuture<Void> putFuture = table.putAsync(message.getKey(), message.getValue().getDefault());
+            return putFuture
+                .exceptionally(ex -> {
+                  LOG.warn("PUT default failed due to an exception. Ignoring the exception and proceeding with update. "
+                          + "The exception encountered is: ", ex);
+                  return null;
+                })
+                .thenCompose(res -> table.updateAsync(message.getKey(), message.getValue().getUpdate()));
```

Comment:
what happens if this fails?

Reply:
Added an exception to indicate that update after put default failed.